File: svc-data/src/main/java/com/beancounter/marketdata/service/MarketDataService.java
package com.beancounter.marketdata.service;

import com.beancounter.common.contracts.PriceRequest;
import com.beancounter.common.contracts.PriceResponse;
import com.beancounter.common.input.AssetInput;
import com.beancounter.common.model.Asset;
import com.beancounter.common.model.MarketData;
import com.beancounter.marketdata.providers.PriceService;
import com.beancounter.marketdata.providers.ProviderUtils;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.Future;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Async;
import org.springframework.scheduling.annotation.AsyncResult;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Service container for MarketData information.
 *
 * @author mikeh
 * @since 2019-01-28
 */
@Service
@Slf4j
public class MarketDataService {

  private final ProviderUtils providerUtils;
  private final PriceService priceService;

  @Autowired
  MarketDataService(ProviderUtils providerUtils,
                    PriceService priceService) {
    this.providerUtils = providerUtils;
    this.priceService = priceService;
  }

  @Transactional
  public void backFill(Asset asset) {
    Collection<Asset> assets = new ArrayList<>();
    assets.add(asset);
    Map<MarketDataProvider, Collection<Asset>>  byFactory =
        providerUtils.splitProviders(providerUtils.getInputs(assets));

    for (MarketDataProvider marketDataProvider : byFactory.keySet()) {
      priceService.process(marketDataProvider.backFill(asset));
    }
  }

  @SneakyThrows
  public PriceResponse getPriceResponse(Asset asset) {
    return getFuturePriceResponse(asset).get();
  }

  /**
   * Prices for the request.
   *
   * @param priceRequest to process
   * @return results
   */
  @Transactional
  public PriceResponse getPriceResponse(PriceRequest priceRequest) {

    Map<MarketDataProvider, Collection<Asset>>
        byFactory = providerUtils.splitProviders(priceRequest.getAssets());

    Collection<MarketData> existing = new ArrayList<>();
    Collection<MarketData> apiResults = new ArrayList<>();

    for (MarketDataProvider marketDataProvider : byFactory.keySet()) {
      // Pull from the DB
      Iterator<Asset> assetIterable = byFactory.get(marketDataProvider).iterator();
      while (assetIterable.hasNext()) {
        Asset asset = assetIterable.next();
        LocalDate mpDate = marketDataProvider.getDate(asset.getMarket(), priceRequest);
        Optional<MarketData> md = priceService.getMarketData(asset.getId(), mpDate);
        if (md.isPresent()) {
          MarketData mdValue = md.get();
          mdValue.setAsset(asset);
          existing.add(mdValue);
          assetIterable.remove(); // One less external query to make
        }
      }

      // Pull the balance over external API integration
      Collection<Asset> apiAssets = byFactory.get(marketDataProvider);
      if (!apiAssets.isEmpty()) {
        Collection<AssetInput> assetInputs = providerUtils.getInputs(apiAssets);
        PriceRequest apiRequest = PriceRequest.builder()
            .date(priceRequest.getDate())
            .assets(assetInputs)
            .build();
        apiResults = marketDataProvider.getMarketData(apiRequest);
      }
    }
    // Merge results into a response
    PriceResponse response = PriceResponse.builder().data(apiResults).build();
    priceService.write(response); // Async write
    response.getData().addAll(existing);
    return response;
  }

  /**
   * Get the current MarketData values for the supplied Asset.
   *
   * @param asset to query
   * @return MarketData - Values will be ZERO if not found or an integration problem occurs
   */
  @Async
  public Future<PriceResponse> getFuturePriceResponse(Asset asset) {
    List<AssetInput> inputs = new ArrayList<>();
    inputs.add(AssetInput.builder()
        .resolvedAsset(asset).build());
    return new AsyncResult<>(getPriceResponse(PriceRequest.builder().assets(inputs).build()));
  }

  /**
   * Delete all prices.  Supports testing
   */
  public void purge() {
    priceService.purge();
  }
}

File: svc-data/src/main/kotlin/com/beancounter/marketdata/service/MarketDataService.kt
package com.beancounter.marketdata.service

import com.beancounter.common.contracts.PriceRequest
import com.beancounter.common.contracts.PriceResponse
import com.beancounter.common.input.AssetInput
import com.beancounter.common.model.Asset
import com.beancounter.common.model.Market
import com.beancounter.common.model.MarketData
import com.beancounter.marketdata.providers.PriceService
import com.beancounter.marketdata.providers.ProviderUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate

/**
 * Service container for MarketData information.
 *
 * @author mikeh
 * @since 2019-01-28
 */
@Service
class MarketDataService @Autowired internal constructor(
    private val providerUtils: ProviderUtils,
    private val priceService: PriceService
) {
    @Transactional
    fun backFill(asset: Asset) {
        val assets: MutableCollection<Asset> = ArrayList()
        assets.add(asset)
        val byFactory = providerUtils.splitProviders(providerUtils.getInputs(assets))
        for (marketDataProvider in byFactory.keys) {
            priceService.process(marketDataProvider.backFill(asset))
        }
    }

    fun getPriceResponse(assetInput: AssetInput): PriceResponse {
        return getPriceResponse(PriceRequest.of(assetInput))
    }

    /**
     * Prices for the request.
     *
     * @param priceRequest to process
     * @return results
     */
    @Transactional
    fun getPriceResponse(priceRequest: PriceRequest): PriceResponse {
        val byFactory = providerUtils.splitProviders(priceRequest.assets)
        val fromDb: MutableCollection<MarketData> = ArrayList()
        var apiResults: Collection<MarketData> = ArrayList()
        var marketDate: LocalDate? = null
        var lastMarket: Market? = null
        for (marketDataProvider in byFactory.keys) {
            // Pull from the DB
            val assetIterable = (byFactory[marketDataProvider] ?: error("")).iterator()
            while (assetIterable.hasNext()) {
                val asset = assetIterable.next()
                if (lastMarket == null || lastMarket.code != asset.market.code) {
                    lastMarket = asset.market
                    marketDate = marketDataProvider.getDate(lastMarket, priceRequest)
                    log.debug("Requested date ${priceRequest.date} resolved as $marketDate")
                }

                val md = priceService.getMarketData(asset.id, marketDate)
                if (md.isPresent) {
                    val mdValue = md.get()
                    mdValue.asset = asset
                    fromDb.add(mdValue)
                    assetIterable.remove() // One less external query to make
                }
            }

            // Pull the balance over external API integration
            val apiAssets = byFactory[marketDataProvider]
            if (!apiAssets!!.isEmpty()) {
                val assetInputs = providerUtils.getInputs(apiAssets)
                val apiRequest = PriceRequest(priceRequest.date, assetInputs)
                apiResults = marketDataProvider.getMarketData(apiRequest)
            }
        }
        // Merge results into a response
        if (fromDb.size + apiResults.size > 1) {
            log.debug("From DB: ${fromDb.size}, from API: ${apiResults.size}")
        }
        if (apiResults.isNotEmpty()) {
            priceService.write(PriceResponse(apiResults)) // Async write
        }
        fromDb.addAll(apiResults)
        return PriceResponse(fromDb)
    }

    /**
     * Delete all prices.  Supports testing
     */
    fun purge() {
        priceService.purge()
    }

    companion object {
        private val log = LoggerFactory.getLogger(MarketDataService::class.java)
    }
}
